File: card_creator.php
# card_creator.php
<?php
$templates = [
    1 => ['bg' => '#FFFFFF', 'text' => '#000000'],
    2 => ['bg' => '#FFD700', 'text' => '#8B0000'],
    3 => ['bg' => '#ADD8E6', 'text' => '#FF4500']
];

function hexToRgb($hex) {
    $hex = str_replace('#', '', $hex);
    if (strlen($hex) == 6) {
        return [hexdec($hex[0].$hex[1]), hexdec($hex[2].$hex[3]), hexdec($hex[4].$hex[5])];
    }
    return [0,0,0];
}

function createCard($name, $age, $message, $template, $bgHex, $textHex, $photoPath, $bgImagePath, $output) {
    $width = 800; $height = 600;
    $img = imagecreatetruecolor($width, $height);

    // Background
    if ($bgImagePath && file_exists($bgImagePath)) {
        $bgImg = imagecreatefromstring(file_get_contents($bgImagePath));
        if ($bgImg) {
            imagecopyresampled($img, $bgImg, 0,0,0,0, $width, $height, imagesx($bgImg), imagesy($bgImg));
            imagedestroy($bgImg);
        } else {
            $bg = $templates[$template]['bg'];
        }
    } else {
        $bg = $bgHex ?: $templates[$template]['bg'];
    }
    if (!isset($bgImg) || !$bgImg) {
        list($r,$g,$b) = hexToRgb($bg);
        $bgColor = imagecolorallocate($img, $r, $g, $b);
        imagefilledrectangle($img, 0, 0, $width, $height, $bgColor);
    }

    // Text color
    $text = $textHex ?: $templates[$template]['text'];
    list($r,$g,$b) = hexToRgb($text);
    $textColor = imagecolorallocate($img, $r, $g, $b);

    // Font paths (using built-in GD fonts for simplicity, but we can use TTF if available)
    $font = '/usr/share/fonts/truetype/liberation/LiberationSans-Bold.ttf';
    $fontReg = '/usr/share/fonts/truetype/liberation/LiberationSans-Regular.ttf';
    $fontItalic = '/usr/share/fonts/truetype/liberation/LiberationSans-Italic.ttf';

    // Title
    if (file_exists($font)) {
        imagettftext($img, 60, 0, 200, 80, $textColor, $font, '🎉 Happy Birthday! 🎉');
        imagettftext($img, 70, 0, 200, 200, $textColor, $font, $name);
        if ($age) imagettftext($img, 50, 0, 200, 280, $textColor, $fontReg, $age);
        // Message wrap
        $wrapped = wordwrap($message, 30, "\n");
        imagettftext($img, 30, 0, 200, 380, $textColor, $fontItalic, $wrapped);
        imagettftext($img, 40, 0, 200, 520, $textColor, $fontReg, '🎂  🎈  🎁');
    } else {
        // Fallback to built-in
        imagestring($img, 5, 200, 50, 'Happy Birthday!', $textColor);
        imagestring($img, 5, 200, 100, $name, $textColor);
        if ($age) imagestring($img, 4, 200, 140, $age, $textColor);
        imagestring($img, 3, 200, 200, $message, $textColor);
    }

    // Photo
    if ($photoPath && file_exists($photoPath)) {
        $photo = imagecreatefromstring(file_get_contents($photoPath));
        if ($photo) {
            $size = 150;
            imagecopyresampled($img, $photo, $width - $size - 20, 20, 0,0, $size, $size, imagesx($photo), imagesy($photo));
            imagedestroy($photo);
        }
    }

    imagepng($img, $output);
    imagedestroy($img);
    echo "Card saved to $output\n";
}

$opts = getopt("n:a:m:t:o:", ["name:", "age:", "message:", "template:", "bg:", "text:", "photo:", "bg-image:", "output:", "html:"]);
$name = $opts['n'] ?? $opts['name'] ?? null;
$age = isset($opts['a']) ? (int)$opts['a'] : (isset($opts['age']) ? (int)$opts['age'] : null);
$message = $opts['m'] ?? $opts['message'] ?? null;
$template = isset($opts['t']) ? (int)$opts['t'] : (isset($opts['template']) ? (int)$opts['template'] : 1);
$bg = $opts['bg'] ?? null;
$text = $opts['text'] ?? null;
$photo = $opts['photo'] ?? null;
$bgImage = $opts['bg-image'] ?? null;
$output = $opts['o'] ?? $opts['output'] ?? 'card.png';
$html = $opts['html'] ?? null;

if (!$name || !$message) {
    fwrite(STDERR, "Error: -n and -m are required\n");
    exit(1);
}
if ($template < 1 || $template > 3) {
    fwrite(STDERR, "Template must be 1-3\n");
    exit(1);
}

createCard($name, $age, $message, $template, $bg, $text, $photo, $bgImage, $output);

if ($html) {
    $htmlContent = <<<HTML
<!DOCTYPE html>
<html><head><title>Birthday Card</title>
<style>body{font-family:sans-serif;text-align:center;background:#f0f0f0;}
.card{background:{$bg};color:{$text};border-radius:20px;padding:40px;max-width:600px;margin:50px auto;box-shadow:0 4px 8px rgba(0,0,0,0.2);}
h1{font-size:3em;}.name{font-size:2.5em;}.age{font-size:2em;}.msg{font-size:1.5em;}</style>
</head><body><div class="card">
<h1>🎉 Happy Birthday! 🎉</h1>
<div class="name">{$name}</div>
HTML;
    if ($age) $htmlContent .= "<div class=\"age\">$age</div>";
    $htmlContent .= "<div class=\"msg\">$message</div>";
    $htmlContent .= '<div style="font-size:3em;">🎂 🎈 🎁</div></div></body></html>';
    file_put_contents($html, $htmlContent);
    echo "HTML card saved to $html\n";
}
?>

File: card_creator.rb
# card_creator.rb
require 'optparse'
require 'rmagick'
include Magick

TEMPLATES = {
  1 => { bg: 'white', text: 'black' },
  2 => { bg: '#FFD700', text: '#8B0000' },
  3 => { bg: '#ADD8E6', text: '#FF4500' }
}

def hex_to_rgb(hex)
  hex = hex[1..-1] if hex.start_with?('#')
  hex.scan(/../).map { |c| c.to_i(16) }
end

def create_card(name, age, message, template, bg_hex, text_hex, photo_path, bg_image_path, output)
  width, height = 800, 600
  bg_color = bg_hex || TEMPLATES[template][:bg]
  text_color = text_hex || TEMPLATES[template][:text]

  # Create image
  if bg_image_path && File.exist?(bg_image_path)
    img = ImageList.new(bg_image_path).first
    img.resize!(width, height)
  else
    bg_rgb = bg_color.is_a?(String) ? hex_to_rgb(bg_color) : [255,255,255]
    img = Image.new(width, height) { self.background_color = "rgb(#{bg_rgb.join(',')})" }
  end

  draw = Draw.new
  text_rgb = text_color.is_a?(String) ? hex_to_rgb(text_color) : [0,0,0]
  draw.fill = "rgb(#{text_rgb.join(',')})"
  draw.gravity = NorthGravity

  # Title
  draw.font = '/usr/share/fonts/truetype/liberation/LiberationSans-Bold.ttf'
  draw.pointsize = 60
  draw.annotate(img, 0,0,0,20, '🎉 Happy Birthday! 🎉')

  # Name
  draw.pointsize = 70
  draw.annotate(img, 0,0,0,140, name)

  # Age
  if age
    draw.pointsize = 50
    draw.annotate(img, 0,0,0,220, age.to_s)
  end

  # Message wrapped
  draw.pointsize = 30
  draw.font = '/usr/share/fonts/truetype/liberation/LiberationSans-Italic.ttf'
  wrapped = message.scan(/.{1,30}/).join("\n")
  draw.annotate(img, 0,0,0,320, wrapped)

  # Emojis
  draw.pointsize = 40
  draw.font = '/usr/share/fonts/truetype/liberation/LiberationSans-Regular.ttf'
  draw.annotate(img, 0,0,0,460, '🎂  🎈  🎁')

  # Photo
  if photo_path && File.exist?(photo_path)
    photo = ImageList.new(photo_path).first
    photo.resize!(150, 150)
    img.composite!(photo, width - photo.columns - 20, 20, Magick::OverCompositeOp)
  end

  img.write(output)
  puts "Card saved to #{output}"
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: card_creator.rb [options]"
  opts.on('-n NAME', '--name NAME', 'Recipient name') { |v| options[:name] = v }
  opts.on('-a AGE', '--age AGE', Integer, 'Age') { |v| options[:age] = v }
  opts.on('-m MESSAGE', '--message MESSAGE', 'Custom message') { |v| options[:message] = v }
  opts.on('-t TEMPLATE', '--template TEMPLATE', Integer, 'Template (1-3)') { |v| options[:template] = v }
  opts.on('--bg COLOR', 'Background color') { |v| options[:bg] = v }
  opts.on('--text COLOR', 'Text color') { |v| options[:text] = v }
  opts.on('--photo PATH', 'Photo path') { |v| options[:photo] = v }
  opts.on('--bg-image PATH', 'Background image path') { |v| options[:bg_image] = v }
  opts.on('-o OUTPUT', '--output OUTPUT', 'Output PNG') { |v| options[:output] = v }
  opts.on('--html HTML', 'Output HTML file') { |v| options[:html] = v }
end.parse!

unless options[:name] && options[:message]
  warn "Error: -n and -m are required"
  exit 1
end
options[:template] ||= 1
options[:output] ||= 'card.png'

create_card(
  options[:name], options[:age], options[:message], options[:template],
  options[:bg], options[:text], options[:photo], options[:bg_image], options[:output]
)

if options[:html]
  html = <<~HTML
  <!DOCTYPE html>
  <html><head><title>Birthday Card</title>
  <style>body{font-family:sans-serif;text-align:center;background:#f0f0f0;}
  .card{background:#{options[:bg] || 'FFFFFF'};color:#{options[:text] || '000000'};
  border-radius:20px;padding:40px;max-width:600px;margin:50px auto;box-shadow:0 4px 8px rgba(0,0,0,0.2);}
  h1{font-size:3em;}.name{font-size:2.5em;}.age{font-size:2em;}.msg{font-size:1.5em;}</style>
  </head><body><div class="card">
  <h1>🎉 Happy Birthday! 🎉</h1>
  <div class="name">#{options[:name]}</div>
  #{options[:age] ? "<div class=\"age\">#{options[:age]}</div>" : ''}
  <div class="msg">#{options[:message]}</div>
  <div style="font-size:3em;">🎂 🎈 🎁</div>
  </div></body></html>
  HTML
  File.write(options[:html], html)
  puts "HTML card saved to #{options[:html]}"
end
